keps/sig-scheduling/4328-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,359 @@
+
+## Design Details
+
+Introducing a new type of NodeAffinity:
+```go
+type NodeAffinity struct {
+	RequiredDuringSchedulingRequiredDuringExecution *NodeSelector `json:"requiredDuringSchedulingRequiredDuringExecution,omitempty"`
+}
+```
+
+Add a controller called `affinityEviction` 
+
+The controller do the following things:
+
+- Listening to the changes of node labels
+- Iterating over all pods assigned to the node, checks the NodeAffinity field, if `RequiredDuringSchedulingRequiredDuringExecution` exists, checks if the selector still met to the new node label. If not, evict the pod.
+- It will add a DisruptionTarget condition with corresponding reason and message for the pod.
+- It will use client-go to delete the pod directly. 
+- It does not respect your configured PodDisruptionBudget.
+- It does respect terminationGracePeriodSeconds.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+Currently coverages:
+- `k8s.io/kubernetes/pkg/scheduler/framework/plugins/nodeaffinity`: `2023-12-13` - `82.5`
+- `k8s.io/kubernetes/pkg/apis/core/validation`:`2023-12-13` - `83.9`
+- `k8s.io/kubernetes/pkg/controller/affinityeviction`:`2023-12-13` - `0` 
+
+These tests will be added:
+- New tests will be added to `plugins/nodeaffinity` to ensure this plugins handles `RequiredDuringSchedulingRequiredDuringExecution` correctly.
+- New tests will be added to `core/validation` to ensure the validation logic for `RequiredDuringSchedulingRequiredDuringExecution` is corrct.
+- A new directory `controller/affinityEviction` will be added, and new tests will be added to ensure that evict is correctly triggered.
+  
+##### Integration tests
+
+These tests will be added:
+- Test that pods with `RequiredDuringSchedulingRequiredDuringExecution` node affinity are correctly scheduled.
+- Test that pods with `RequiredDuringSchedulingRequiredDuringExecution` node affinity are evicted if the corresponding node's label changed and the selectors are no longer met.
+
+##### e2e tests
+
+These tests will be added:
+- `RequiredDuringSchedulingRequiredDuringExecution` will be tested during both scheduling and execution.
+  1. create a node that has a label
+  2. create a pod matches the label and has `RequiredDuringSchedulingRequiredDuringExecution`
+  3. verify that the pod is correctly scheduled to the node
+  4. remove the node label
+  5. verify that the pod is evicted
+  6. create a new pod with same template
+  7. verify that the pod failed scheduling
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial unit/e2e tests completed and enabled
+- Documentation is added to demonstrate why this is useful how exactly affinity needs to be configured.
+
+#### Beta
```

Comment:
So in beta, you will probably add e2e tests and metrics. I would mention the metric additions here.